### PR TITLE
Add JSONParser

### LINKED ISSUE #6: JSON config parser?
It would be pretty trivial to write. If I open a PR, would you accept it, or would you prefer people write their own config parsers?

### PR DESCRIPTION
**Pull request by carlmjohnson**
Fixes #6.

### CODE REVIEW

File: parse.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ func Parse(fs *flag.FlagSet, args []string, options ...Option) error {
 		}
 		defer f.Close()
 
-		c.configFileParser(f, func(name, value string) error {
+		err = c.configFileParser(f, func(name, value string) error {
```
Comment (peterbourgon) [inline parse.go]: Ah, thanks.

File: json_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			name: "Key with array of values",
+			args: []string{},
+			file: `
+                {
+                    "s": ["t", "s"],
+                    "i": ["11", "10"],
+                    "b": [false, true],
+                    "d": ["10m", "5s"]
+                }
+            `,
```
Comment (peterbourgon) [inline json_test.go]: Nit: can you please fix this up with tabs for indention?

File: json_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			want: want{"s", 10, true, 5 * time.Second, ""},
+		},
+		{
+			name: "bad JSON file",
+			args: []string{},
+			file: `{`,
+			want: want{d: time.Second, err: "error parsing JSON config"},
```
Comment (peterbourgon) [inline json_test.go]: Is there a way to improve this, to leverage the ParseError type, so you're not comparing strings?

File: json.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+// multiple times.
+func JSONParser(r io.Reader, set func(name, value string) error) error {
+	var m map[string]interface{}
+	d := json.NewDecoder(r)
+	// Must set UseNumber for stringifyValue to work
+	d.UseNumber()
+	err := d.Decode(&m)
+	if err != nil {
+		return errors.Wrap(err, "error parsing JSON config")
+	}
```
Comment (peterbourgon) [inline json.go]: Nit: just to make this a bit tighter...

```go
d := json.NewDecoder(r)
d.UseNumber() // stringify expects json.Numbers
var m map[string]interface{}
if err := d.Decode(&m); err != nil {
    return ParseError{Err: err}
}
```

File: json.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+func stringifyValue(val interface{}) (string, error) {
+	switch v := val.(type) {
+	case string:
+		return v, nil
+	case json.Number:
+		return v.String(), nil
+	case bool:
+		return strconv.FormatBool(v), nil
+	}
+	return "", errors.Errorf("could not convert %q (type %T) to string", val, val)
```
Comment (peterbourgon) [inline json.go]: Extremely minor nit: can you move this to a `default` clause?

File: json.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
+package ff
```
Comment (peterbourgon) [inline json.go]: Honest question: do you think it's better to move this to its own ffjson subpackage? Pro: symmetry with potential future ffyaml subpackage. Con: this one only uses stdlib, so not strictly necessary?

Comment (carlmjohnson) [reply json.go]: Might make more sense.

File: json_test.go (hunk 3 above)
Comment (carlmjohnson) [reply json_test.go]: Which ParseError? You mean the errors in encoding/json? For a test, I'm inclined to be happy with just a quick and dirty string test.

File: json.go (hunk 4 above)
Comment (carlmjohnson) [reply json.go]: If I'm creating a public facing error type, then yeah, this should probably be in a new package.

**Comment by peterbourgon** [comment]
May end up breaking this out but fine for now I think.

**Pull request merged**